test/framework/kubetest/setup.go

```diff
@@ -0,0 +1,106 @@
+limitations under the License.
+*/
+
+package kubetest
+
+import (
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	ciVersionURL = "https://dl.k8s.io/ci/k8s-master.txt"
+	// E2EVariableUseCIArtifacts if "true" will fetch the latest build from the main branch of kubernetes/kubernetes
+	E2EVariableUseCIArtifacts = "KUBETEST_USE_CI_ARTIFACTS"
+	// E2EVariableConfigFilePath is the path to the kubetest e2e config file
+	E2EVariableConfigFilePath = "KUBETEST_CONFIG_FILE_PATH"
+	// E2EVariableGinkgoNodes is the number of Ginkgo nodes to use
+	E2EVariableGinkgoNodes = "KUBETEST_GINKGO_NODES"
+	// E2EVariableGinkgoSlowSpecThreshold is time in s before spec is marked as slow
+	E2EVariableGinkgoSlowSpecThreshold = "KUBETEST_GINKGO_SLOW_SPEC_THRESHOLD"
+)
+
+// Configuration creates a new kubetest configuration
+type Configuration struct {
+	// UseCIArtifacts will fetch the latest build from the main branch of kubernetes/kubernetes
+	UseCIArtifacts bool `json:"useCIArtifacts,omitempty"`
+	// ArtifactsDirectory is where conformance suite output will go
+	ArtifactsDirectory string `json:"artifactsDirectory,omitempty"`
+	// E2EConfig is the clusterctl E2E config, which will be modified for conformance
+	E2EConfig *clusterctl.E2EConfig `json:"e2eConfig,omitempty"`
+	// Path to the kubetest e2e config file
+	ConfigFilePath string `json:"configFilePath,omitempty"`
+	// GinkgoNodes is the number of Ginkgo nodes to use
+	GinkgoNodes int `json:"ginkgoNodes,omitempty"`
+	// GinkgoSlowSpecThreshold is time in s before spec is marked as slow
+	GinkgoSlowSpecThreshold int `json:"ginkgoSlowSpecThreshold,omitempty"`
+}
+
+// NewConfiguration creates a new kubetest configuration. This may
+// also mutate the clusterctl e2econfig if UseCIArtifacts is enabled.
+func NewConfiguration(e2eConfig *clusterctl.E2EConfig, artifactsDirectory string) *Configuration {
+
+	useCIArtifacts, err := strconv.ParseBool(e2eConfig.GetStringVariableWithDefault(E2EVariableUseCIArtifacts, "false"))
+	Expect(err).NotTo(Occur())
+	configFilePath := e2eConfig.GetVariable(E2EVariableConfigFilePath)
+	ginkgoNodes, err := strconv.ParseInt(e2eConfig.GetStringVariableWithDefault(E2EVariableGinkgoNodes, "1"))
+	ginkgoSlowSpecThreshold, err := strconv.ParseInt(e2eConfig.GetStringVariableWithDefault(E2EVariableGinkgoSlowSpecThreshold), "120")
+	Expect(err).NotTo(Occur())
+
+	c := Configuration{
+		E2EConfig:               e2eConfig,
+		UseCIArtifacts:          useCIArtifacts,
+		ArtifactsDirectory:      artifactsDirectory,
+		ConfigFilePath:          configFilePath,
+		GinkgoNodes:             ginkgoNodes,
+		GinkgoSlowSpecThreshold: ginkgoSlowSpecThreshold,
+	}
+
+	if c.UseCIArtifacts {
+		newVersion := fetchKubernetesCIVersion()
+		c.E2EConfig.Variables["KUBERNETES_VERSION"] = newVersion
+		c.E2EConfig.Variables["CI_VERSION"] = newVersion
```

Comment:
I think that having something in Kubetest that alters E2EConfig variables can be confusing.
What about having something in `LoadE2EConfig` that automatically resolves the Kubernetes version? If the Kubernetes version is a semantic version, we are ok, if it is a URL (or a well know tag like master), it resolves it.
This will also avoid the needs of the KUBETEST_USE_CI_ARTIFACTS variable.

Also CI_VERSION seems not used elsewhere, is it possible to remove?